server/src/completions.ts

```diff
@@ -0,0 +1,414 @@
+import {
+  CompletionItem,
+  CompletionList,
+  CompletionItemKind,
+  Position,
+} from 'vscode-languageserver'
+import { TextDocument } from 'vscode-languageserver-textdocument'
+import { MyBlock } from './MessageHandler'
+import {
+  blockAttributes,
+  fieldAttributes,
+  allowedBlockTypes,
+  corePrimitiveTypes,
+  supportedDataSourceFields,
+  supportedGeneratorFields,
+} from './completionUtil'
+
+
+function toCompletionItems(
+  allowedTypes: string[],
+  kind: CompletionItemKind,
+): CompletionItem[] {
+  return allowedTypes.map((label) => ({ label, kind }))
+}
+
+function getSuggestionForBlockAttribute(blockType: string, document: TextDocument, position: Position): CompletionItem[] {
+  if (blockType !== 'model') {
+    return []
+  }
+
+  const symbolsBeforePosition = document.getText({
+    start: { line: position.line, character: position.character - 2 },
+    end: { line: position.line, character: position.character },
+  })
+  const symbolBeforePosition = symbolsBeforePosition.charAt(1)
+  // create deep copy
+  const suggestions: CompletionItem[] = JSON.parse(JSON.stringify(blockAttributes))
```

Comment:
You can use `klona` instead, which is significantly faster (10x).
https://github.com/lukeed/klona